File: src/core/spec.ts
/**
 * The wire shape.
 *
 * Components live in a flat map keyed by id, with a named root. They are not a
 * nested tree. Google's A2UI and Vercel's json-render were built independently
 * and both landed here, for the same reason: a nested tree cannot be patched or
 * streamed out of order, and a flat map can. A child may arrive before its
 * parent, a parent before its children, and neither case needs special handling.
 *
 * Data lives separately from components, addressed by JSON Pointer (RFC 6901),
 * so a value can change without touching the component graph and a component can
 * be replaced without losing its value.
 */

/** A component id. One path segment: letters, digits, `_`, `-`. */
export type ComponentId = string;

/** A JSON Pointer into the surface data model, e.g. `/report/rows/0/total`. */
export type Pointer = string;

/** Any value the model can put in a prop or the data model. */
export type Json =
  | string
  | number
  | boolean
  | null
  | Json[]
  | { [key: string]: Json };

/**
 * A prop value that reads from the data model instead of being literal.
 * Input components write back through the same pointer.
 */
export interface Binding {
  $bind: Pointer;
}

export type PropValue = Json | Binding | Computed;

export function isBinding(v: unknown): v is Binding {
  return (
    typeof v === "object" &&
    v !== null &&
    "$bind" in v &&
    typeof (v as Binding).$bind === "string"
  );
}

/**
 * A prop derived from the data rather than read out of it.
 *
 * Three forms, and each one earned its place by a model reaching for it. An app
 * that tracks anything wants "how many" and "how much"; the first live run
 * against a real model added "average rating" to a book tracker unprompted, so
 * $avg is here too.
 *
 * Every additional operator is a small expression language that has to be
 * authored correctly by a model, understood by a reader, and kept safe, so the
 * bar for a fourth is watching a model want it. Anything more complicated
 * belongs in the data.
 */
export type Computed =
  | { $count: Pointer; where?: { field: string; equals: Json } }
  | { $sum: Pointer; field: string }
  | { $avg: Pointer; field: string };

export function isComputed(v: unknown): v is Computed {
  if (typeof v !== "object" || v === null) return false;
  return "$count" in v || "$sum" in v || "$avg" in v;
}

export interface ComponentNode {
  id: ComponentId;
  /** Component name, must exist in the catalog. */
  type: string;
  props: Record<string, PropValue>;
  /** Ordered child ids. May reference ids that have not arrived yet. */
  children: ComponentId[];
}

/** A complete surface at a point in time. */
export interface Spec {
  root: ComponentId | null;
  elements: Record<ComponentId, ComponentNode>;
  data: Record<string, Json>;
}

export function emptySpec(): Spec {
  return { root: null, elements: {}, data: {} };
}

/* -------------------------------------------------------------------------- */
/* Stream operations                                                          */
/* -------------------------------------------------------------------------- */

/**
 * The four things a model can say. Deliberately small: every generative UI
 * protocol surveyed reduces to these, and a fifth op is usually a sign the
 * catalog should have absorbed the behaviour instead.
 */
export type Op =
  | { op: "component"; node: ComponentNode }
  | { op: "children"; id: ComponentId; children: ComponentId[] }
  | { op: "data"; path: Pointer; value: Json | undefined }
  | { op: "root"; id: ComponentId };

/** Emitted by the store as a surface assembles. */
export type SurfaceEvent =
  /** Root arrived and resolves. Nothing should render before this. */
  | { type: "ready"; spec: Spec }
  /** The graph or data changed after ready. */
  | { type: "patch"; spec: Spec; changed: ComponentId[] }
  /** A child id was referenced that has not arrived. Render a placeholder. */
  | { type: "pending"; ids: ComponentId[] }
  /** The stream ended cleanly. */
  | { type: "done"; spec: Spec }
  /** Recoverable problem. The surface keeps going. */
  | { type: "warn"; message: string; detail?: unknown }
  /** Unrecoverable. The surface is abandoned. */
  | { type: "error"; message: string; detail?: unknown };

File: hud/Sources/BobHUDKit/Spec.swift
import Foundation

/// The wire shape, ported from `src/core/spec.ts`.
///
/// Components live in a flat map keyed by id with a named root, not a nested
/// tree. That is what makes a stream renderable out of order: a child may arrive
/// before its parent and neither case needs special handling. Google's A2UI and
/// Vercel's json-render landed on the same shape independently.

public typealias ComponentID = String

/// A JSON value. Swift has no native one, and pulling a dependency for six cases
/// is not worth it in a process whose whole point is starting instantly.
public enum JSON: Sendable, Equatable {
    case string(String)
    case number(Double)
    case bool(Bool)
    case null
    case array([JSON])
    case object([String: JSON])

    public var stringValue: String? {
        switch self {
        case .string(let s): return s
        case .number(let n): return n == n.rounded() ? String(Int(n)) : String(n)
        case .bool(let b): return b ? "yes" : "no"
        case .null: return nil
        default: return nil
        }
    }

    public var doubleValue: Double? {
        switch self {
        case .number(let n): return n
        case .string(let s): return Double(s)
        case .bool(let b): return b ? 1 : 0
        default: return nil
        }
    }

    public var arrayValue: [JSON]? {
        if case .array(let a) = self { return a }
        return nil
    }

    public var objectValue: [String: JSON]? {
        if case .object(let o) = self { return o }
        return nil
    }

    /// Display text for a table cell or a label.
    public var display: String {
        switch self {
        case .string(let s): return s
        case .number(let n): return n == n.rounded() && abs(n) < 1e15 ? String(Int(n)) : String(n)
        case .bool(let b): return b ? "yes" : "no"
        case .null: return ""
        case .array(let a): return a.map(\.display).joined(separator: ", ")
        case .object: return ""
        }
    }
}

/// A prop that reads from the data model rather than carrying a literal.
public struct Binding: Sendable, Equatable {
    public let pointer: String
    public init(pointer: String) { self.pointer = pointer }
}

/// A prop derived from the data. The same three forms the TypeScript supports,
/// and no more: every operator is a small expression language that has to be
/// authored correctly by a model and understood by whoever reads the output.
public enum Computed: Sendable, Equatable {
    case count(path: String, whereField: String?, equals: JSON?)
    case sum(path: String, field: String)
    case avg(path: String, field: String)
}

public enum PropValue: Sendable, Equatable {
    case literal(JSON)
    case binding(Binding)
    case computed(Computed)
}

public struct ComponentNode: Sendable, Equatable {
    public var id: ComponentID
    public var type: String
    public var props: [String: PropValue]
    public var children: [ComponentID]

    public init(
        id: ComponentID, type: String,
        props: [String: PropValue] = [:], children: [ComponentID] = []
    ) {
        self.id = id
        self.type = type
        self.props = props
        self.children = children
    }

    /// Marks an edge held for a parent that has not arrived yet.
    public static let pendingType = "__pending__"
}

public struct Spec: Sendable, Equatable {
    public var root: ComponentID?
    public var elements: [ComponentID: ComponentNode]
    public var data: [String: JSON]

    public init(
        root: ComponentID? = nil,
        elements: [ComponentID: ComponentNode] = [:],
        data: [String: JSON] = [:]
    ) {
        self.root = root
        self.elements = elements
        self.data = data
    }
}

/// The four things a stream can say.
public enum Op: Sendable, Equatable {
    case component(ComponentNode)
    case children(id: ComponentID, children: [ComponentID])
    case data(path: String, value: JSON?)
    case root(id: ComponentID)
}
